lib/IDE/SyntaxModel.cpp

```diff
@@ -1229,7 +1229,7 @@ bool ModelASTWalker::handleSpecialDeclAttribute(const DeclAttribute *D,
         assert(0 && "No TokenNodes?");
     }
     if (!passTokenNodesUntil(D->getRange().End,
-                             IncludeNodeAtLocation).shouldContinue)
+                             ExcludeNodeAtLocation).shouldContinue)
```

Comment:
My understanding of `SourceRange` is that the end loc should point to the start of the last token in the range (vs `CharSourceRange` were the end loc is the past-the-end position). I'm guessing the parser is setting the end loc as the start of the `)` when it's there and the start of whatever token it saw instead of the `)` when it's missing (`@` in the test case?) rather than the start of last token it processed (`10.12`?).  If so, fixing the loc the parser passes would be better so that `Lexer::getCharSourceRangeFromSourceRange` still work correctly with the source ranges returned from invalid attributes (fixits and a bunch of our stuff use that method), but feel free to leave it if it's a pain to do. The current fix resolves the crash and shouldn't cause any problems. The last ')' that we would now miss in the valid case should still get passed when we next call `passTokenNodesUntil` with the start location of whatever the next node is. It would be good to add a comment here explaining why we exclude it though.

Reply:
Ah, of course you are right. I have worked too much with `CharSourceRange` recently and mixed the two up.